src/transport_portals4.c

```diff
@@ -120,25 +115,184 @@ static ptl_pt_index_t heap_pt = PTL_PT_ANY;
 #endif
 
 #ifdef ENABLE_THREADS
+shmem_internal_mutex_t shmem_internal_mutex_ptl4_ctx;
 shmem_internal_mutex_t shmem_internal_mutex_ptl4_pt_state;
 shmem_internal_mutex_t shmem_internal_mutex_ptl4_frag;
 shmem_internal_mutex_t shmem_internal_mutex_ptl4_event_slots;
 shmem_internal_mutex_t shmem_internal_mutex_ptl4_nb_fence;
 #endif
 
+static shmem_transport_ctx_t** shmem_transport_portals4_contexts = NULL;
+static size_t shmem_transport_portals4_contexts_len = 0;
+static size_t shmem_transport_portals4_grow_size = 128;
+
+#define SHMEM_TRANSPORT_CTX_DEFAULT_ID -1
 shmem_transport_ctx_t shmem_transport_ctx_default;
 shmem_ctx_t SHMEM_CTX_DEFAULT = (shmem_ctx_t) &shmem_transport_ctx_default;
 
-int shmem_transport_ctx_create(long options, shmem_transport_ctx_t **ctx) {
-  *ctx = NULL;
-  return 0;
+static int
+shmem_transport_ctx_init(shmem_transport_ctx_t *ctx, long options, int id)
+{
+    int ret;
+    ptl_md_t md;
+
+    ctx->id = id;
+    ctx->options = options;
+    ctx->put_md = PTL_INVALID_HANDLE;
+    ctx->get_md = PTL_INVALID_HANDLE;
+    ctx->put_ct = PTL_INVALID_HANDLE;
+    ctx->get_ct = PTL_INVALID_HANDLE;
+    shmem_internal_atomic_write(&ctx->pending_put_cntr, 0);
+    shmem_internal_atomic_write(&ctx->pending_get_cntr, 0);
+
+    /* Allocate put completion tracking resources */
+    ret = PtlCTAlloc(shmem_transport_portals4_ni_h, &ctx->put_ct);
+    if (PTL_OK != ret) {
+        RAISE_WARN_MSG("Put CT allocation failed: %d\n", ret);
+        goto cleanup;
+    }
+
+    md.start = 0;
+    md.length = PTL_SIZE_MAX;
+    md.options = PTL_MD_EVENT_CT_ACK | PTL_MD_EVENT_SUCCESS_DISABLE;
+    if (1 != PORTALS4_TOTAL_DATA_ORDERING) {
+        md.options |= PTL_MD_UNORDERED;
+    }
+    md.eq_handle = shmem_transport_portals4_eq_h;
+    md.ct_handle = ctx->put_ct;
+    ret = PtlMDBind(shmem_transport_portals4_ni_h, &md, &ctx->put_md);
+    if (PTL_OK != ret) {
+        RETURN_ERROR_MSG("PtlMDBind of put MD failed: %d\n", ret);
+        goto cleanup;
```

Comment:
This is a fairly common idiom for this type of "try - catch" in C.  What's nice about structuring the code in this way, is it allows us to do cleanup for resources whose references are only in-scope within this function's body.  It's not the case right now, but could be in the future.  It also keeps all of the code within the same function -- if we pulled this out into a cleanup routine, it would be callable from elsewhere, but probably shouldn't be used except for as cleanup for this routine.

I'm mostly ambivalent, but I think I slightly prefer the goto approach since it is more self-contained and results in less code.  We do have precedent for this type of structure in the Portals ```shmem_transport_startup```.  Anyway, happy to hear arguments in favor of the function approach.

Reply:
I'm mostly ambivalent as well, just thought I'd mention it since `goto` is so infamous.  I also think this pattern is actually a good use-case for `goto`.